eventhubs-streamanalytics-eventhubs/tools/eh-asa-perfmon/Program.cs

```diff
@@ -82,69 +79,66 @@ public Task ProcessEventsAsync(IEnumerable<EventData> events)
-        {
-            this._client = EventHubClient.CreateFromConnectionString(connectionString);
-
-            this._cancellationToken = _cancellationToken;
-        }
-
-        public async Task Receive()
+        public async Task Receive(string connectionString, CancellationTokenSource cancellationSource)
         {
             Console.WriteLine("The application will now start to listen for incoming message.");
 
             var csvOutput = File.CreateText("./result.csv");
-            csvOutput.WriteLine("PartitionId,EventCount,BatchCount,BatchFrom,BatchTo,MinLatency,MaxLatency,AvgLatency");            
-
-            var runtimeInfo = await _client.GetRuntimeInformationAsync();
-            Console.WriteLine("Creating receiver handlers...");
-            var utcNow = DateTime.UtcNow;
-            var receivers = runtimeInfo.PartitionIds
-                .Select(pid => {
-                    var receiver = _client.CreateReceiver("$Default", pid, EventPosition.FromEnqueuedTime(utcNow));
-                    Console.WriteLine("Created receiver for partition '{0}'.", pid);
-                    receiver.SetReceiveHandler(new MyPartitionReceiver(pid, csvOutput));
-                    return receiver;
-                })
-                .ToList();
+            csvOutput.WriteLine("PartitionId,EventCount,BatchCount,BatchFrom,BatchTo,MinLatency,MaxLatency,AvgLatency");
 
-            try
-            {
-                await Task.Delay(-1, this._cancellationToken);
-            }
-            catch (TaskCanceledException)
+            await using (var consumerClient = new EventHubConsumerClient(EventHubConsumerClient.DefaultConsumerGroupName, connectionString))
             {
-                // This is okay because the task was cancelled. :)
-            }
-            finally
-            {
-                // Clean up nicely.
-                await Task.WhenAll(
-                    receivers.Select(receiver => receiver.CloseAsync())
-                );
-            }
+                string[] partitions = await consumerClient.GetPartitionIdsAsync();
+
+                ReadEventOptions readOptions = new ReadEventOptions
+                {
+                    MaximumWaitTime = TimeSpan.FromMilliseconds(150)
+                };
+
+                cancellationSource.CancelAfter(TimeSpan.FromSeconds(60));
+
+                List<EventData> receivedEvents = new List<EventData>();
+
+                foreach (string partitionId in partitions)
+                {
+                    var myPartitionReceiver = new MyPartitionReceiver(partitionId, csvOutput);
+                    int eventBatchSize = myPartitionReceiver.MaxBatchSize;
+
+                    Console.WriteLine("Receive events from partion '{0}'.", partitionId);
+
+                    await foreach (PartitionEvent currentEvent in consumerClient.ReadEventsFromPartitionAsync(partitionId, EventPosition.FromEnqueuedTime(DateTime.UtcNow), readOptions, cancellationSource.Token))
```

Comment:
Because this is not a performance-sensitive path, and the previous implementation does not take any steps to guarantee fairness for partition ordering or distribution, I'd recommend moving to `ReadEventsFromPartitions` and eliminate the need to iterate through all partitions.

Reply:
You would lose the ability to set the arbitrary event position using that approach, so you'd probably want to use the beginning of the partition stream, since the templates are creating new resources for this sample.

Reply:
we did not find ReadEventsFromPartitions, refactored with found method ReadEventsAsync.